File: src/algorithms/strings/edit-distance/jaro-winkler-similarity/sources/jaro-winkler-similarity_test.go
package main

import (
	"math"
	"testing"
)

func TestJaroWinklerSimilarityMarthaMarhta(t *testing.T) {
	score := jaroWinklerSimilarity("martha", "marhta")
	if math.Abs(score-0.9611) > 0.0001 {
		t.Errorf("expected ~0.9611, got: %f", score)
	}
}

func TestJaroWinklerSimilarityIdenticalStrings(t *testing.T) {
	if jaroWinklerSimilarity("abc", "abc") != 1.0 {
		t.Error("expected 1.0 for identical strings")
	}
}

func TestJaroWinklerSimilarityTwoEmptyStrings(t *testing.T) {
	if jaroWinklerSimilarity("", "") != 1.0 {
		t.Error("expected 1.0 for two empty strings")
	}
}

func TestJaroWinklerSimilaritySourceEmpty(t *testing.T) {
	if jaroWinklerSimilarity("", "abc") != 0.0 {
		t.Error("expected 0.0 when source is empty")
	}
}

func TestJaroWinklerSimilarityTargetEmpty(t *testing.T) {
	if jaroWinklerSimilarity("abc", "") != 0.0 {
		t.Error("expected 0.0 when target is empty")
	}
}

func TestJaroWinklerSimilarityCompletelyDifferent(t *testing.T) {
	if jaroWinklerSimilarity("abc", "xyz") != 0.0 {
		t.Error("expected 0.0 for completely different strings")
	}
}

func TestJaroWinklerSimilarityCrateTrace(t *testing.T) {
	score := jaroWinklerSimilarity("CRATE", "TRACE")
	if score <= 0.7 || score >= 0.8 {
		t.Errorf("expected score between 0.7 and 0.8, got: %f", score)
	}
}

func TestJaroWinklerSimilarityDwayneDuane(t *testing.T) {
	score := jaroWinklerSimilarity("DwAyNE", "DuANE")
	if score < 0.84 {
		t.Errorf("expected >= 0.84, got: %f", score)
	}
}

func TestJaroWinklerSimilarityIdenticalSingleChars(t *testing.T) {
	if jaroWinklerSimilarity("a", "a") != 1.0 {
		t.Error("expected 1.0 for identical single chars")
	}
}

func TestJaroWinklerSimilarityValueInRange(t *testing.T) {
	score := jaroWinklerSimilarity("algorithm", "logarithm")
	if score < 0.0 || score > 1.0 {
		t.Errorf("expected value between 0.0 and 1.0, got: %f", score)
	}
}

func TestJaroWinklerSimilaritySymmetric(t *testing.T) {
	forward := jaroWinklerSimilarity("martha", "marhta")
	backward := jaroWinklerSimilarity("marhta", "martha")
	if forward != backward {
		t.Errorf("expected symmetric: forward=%f backward=%f", forward, backward)
	}
}

func TestJaroWinklerSimilarityPrefixBonus(t *testing.T) {
	withPrefix := jaroWinklerSimilarity("JOHNSON", "JHNSON")
	withoutPrefix := jaroWinklerSimilarity("AOHNSON", "JHNSON")
	if withPrefix <= withoutPrefix {
		t.Error("expected withPrefix > withoutPrefix")
	}
}

func TestJaroWinklerSimilarityPrefixCappedAtFour(t *testing.T) {
	fourPrefix := jaroWinklerSimilarity("abcdefgh", "abcdXXXX")
	threePrefix := jaroWinklerSimilarity("abcXefgh", "abcdXXXX")
	if fourPrefix <= threePrefix {
		t.Error("expected fourPrefix > threePrefix")
	}
}

File: src/algorithms/strings/edit-distance/jaro-winkler-similarity/__tests__/jaro-winkler-similarity.test.ts
/** Correctness tests for the jaroWinklerSimilarity pure function. */

import { describe, it, expect } from "vitest";
import { jaroWinklerSimilarity } from "../sources/jaro-winkler-similarity.ts?fn";

describe("jaroWinklerSimilarity", () => {
  it('scores "martha" and "marhta" at ~0.9611 (classic example)', () => {
    expect(jaroWinklerSimilarity("martha", "marhta")).toBeCloseTo(0.9611, 4);
  });

  it("returns 1.0 for identical strings", () => {
    expect(jaroWinklerSimilarity("abc", "abc")).toBe(1.0);
  });

  it("returns 1.0 for two empty strings", () => {
    expect(jaroWinklerSimilarity("", "")).toBe(1.0);
  });

  it("returns 0.0 when source is empty", () => {
    expect(jaroWinklerSimilarity("", "abc")).toBe(0.0);
  });

  it("returns 0.0 when target is empty", () => {
    expect(jaroWinklerSimilarity("abc", "")).toBe(0.0);
  });

  it("returns 0.0 for completely different strings of equal length", () => {
    // No characters fall within each other's match windows
    expect(jaroWinklerSimilarity("abc", "xyz")).toBe(0.0);
  });

  it('scores "CRATE" and "TRACE" with partial matches', () => {
    const score = jaroWinklerSimilarity("CRATE", "TRACE");
    // Jaro is ~0.7333, no common prefix so Winkler adds nothing
    expect(score).toBeGreaterThan(0.7);
    expect(score).toBeLessThan(0.8);
  });

  it('scores "DwAyNE" and "DuANE" above 0.84 (prefix boost)', () => {
    const score = jaroWinklerSimilarity("DwAyNE", "DuANE");
    expect(score).toBeGreaterThanOrEqual(0.84);
  });

  it("scores identical single characters at 1.0", () => {
    expect(jaroWinklerSimilarity("a", "a")).toBe(1.0);
  });

  it("returns a value between 0.0 and 1.0 for arbitrary strings", () => {
    const score = jaroWinklerSimilarity("algorithm", "logarithm");
    expect(score).toBeGreaterThanOrEqual(0.0);
    expect(score).toBeLessThanOrEqual(1.0);
  });

  it("is not symmetric — source and target order can differ slightly", () => {
    // By definition Jaro-Winkler is symmetric; confirm both directions are equal
    const forward = jaroWinklerSimilarity("martha", "marhta");
    const backward = jaroWinklerSimilarity("marhta", "martha");
    expect(forward).toBe(backward);
  });

  it("rewards common prefix — longer prefix gives higher score than no prefix", () => {
    // "JOHNSON" vs "JHNSON" shares prefix "J"
    // "AOHNSON" vs "JHNSON" shares no prefix
    const withPrefix = jaroWinklerSimilarity("JOHNSON", "JHNSON");
    const withoutPrefix = jaroWinklerSimilarity("AOHNSON", "JHNSON");
    expect(withPrefix).toBeGreaterThan(withoutPrefix);
  });

  it("prefix bonus is capped at 4 characters", () => {
    // "abcdefgh" vs "abcdXXXX" — prefix length would be 4 (capped)
    // "abcXefgh" vs "abcdXXXX" — prefix length is 3
    const fourPrefixScore = jaroWinklerSimilarity("abcdefgh", "abcdXXXX");
    const threePrefixScore = jaroWinklerSimilarity("abcXefgh", "abcdXXXX");
    expect(fourPrefixScore).toBeGreaterThan(threePrefixScore);
  });
});
